### DISCUSSION TITLE
What's exactly happening while revalidating static pages?

### DISCUSSION THREAD

**Discussion by shreyas-jadhav** [vercel/next.js#23603] [Help] [answered]
I have gone through the docs for [Incremental Static Regeneration](https://nextjs.org/docs/basic-features/data-fetching#incremental-static-regeneration), but I did not find it much sufficient to know what is exactly happening behind the scenes.

>  // Next.js will attempt to re-generate the page:
    // - When a request comes in
    // - At most once every second 
```revalidate: 1,``` // In seconds

Here my doubts are, if I provide ```revalidate: 1```, 

- Does the server constantly run the ```getStaticProps()``` in background every second and fetch data from my database? 

- Does the server re-generate the page every 1 second anyway, or only when the data from the database is new/updated as compared to previous data? Because, re-generating a page every second would cause load on the server and it wouldn't be ideal.

- Does ```revalidate: 1``` mean that I would get **86400 database reads** for every 24 hours (24h = 86400s) ? If so, considering I'm using Firebase Firestore, wouldn't it be more cheaper to establish a connection to listen to documents instead for reading them every second? If that's case how can implement  [```useDocumentData()```](https://github.com/CSFrequency/react-firebase-hooks/tree/master/firestore#usedocumentdata) for listening documents' real-time changes and re-generate the page only when the data changes and not every second constantly?

**Comment by miguderp** [comment]
From my understanding, it only revalidates when the page is requested by a client.

So if no one visits a specific page of yours for a year (but in the meantime you edit its content through a CMS - for example), the first person visiting the page will immediately see the old content and, in the back, trigger a request to revalidate/refetch (because the current visit is older than 1s after the timestamp of that content being generated) and refresh the page’s content for the next person to visit it.

If you are using a tool like [SWR](https://swr.vercel.app/) to revalidate, you can also provide the updated content to the user without having him to refresh the page to see it.

**Comment by sergioviniciuss** [comment, marked as answer]
Regarding your point 3:
No. it's not like that. You're actually saving a lot of requests to the DB due to this feature.
Just complementing what @miguderp said, think of it this way: if your page has 100 users requesting it at the same time (let's say then, per second), and you have set it to revalidate at most at 1s. If this page is fetching data from a db, you'll only be making one request to it, that will build the page again, and you'll be saving 99 requests. The first user triggers the build and the others just benefit from it. (they'll though, get the outdated page while it's not yet generated - this process is called stale while revalidate). 
If no one access the page, then no request is made to the db. This is not a listener for changes, it's actually a mechanism to reduce the amount of requests and automatically build the page for you in case there are changes.

Now, what I really would also like to know is if there's no change in the database, when a user access it will it rebuild the page anyway or just keep the cached version, since nothing has changed. I'm struggling to find out this answer and I thought I'd find it here.

**Reply by sergioviniciuss** [reply]
OK so I played a bit today and found out that pages are always regenerated when a build is triggered, no matter if there's changes or not. I answered it here: https://github.com/vercel/next.js/discussions/25502 just in case. :)
